services/mlmodel/tflite_cpu/tflite_cpu.go

```diff
@@ -0,0 +1,241 @@
+	ModelPath  string  `json:"model_path"`
+	NumThreads int     `json:"num_threads"`
+	LabelPath  *string `json:"label_path"`
+}
+
+// Model is a struct that implements the TensorflowLite CPU implementation of the MLMS.
+// It includes the configured parameters, model struct, and associated metadata.
+type Model struct {
+	attrs    TFLiteConfig
+	model    *inf.TFLiteStruct
+	metadata *mlmodel.MLMetadata
+}
+
+// newTFLiteCPUModel is a constructor that builds a tflite cpu implementation of the MLMS.
+func newTFLiteCPUModel(ctx context.Context, r robot.Robot, conf config.Service, logger golog.Logger) (*Model, error) { //nolint:unparam
+	ctx, span := trace.StartSpan(ctx, "service::mlmodel::NewTFLiteCPUModel")
+	defer span.End()
+
+	// Read ML model service parameters into a TFLiteDetectorConfig
+	var t TFLiteConfig
+	tfParams, err := config.TransformAttributeMapToStruct(&t, conf.Attributes)
+	if err != nil {
+		return &Model{}, errors.Wrapf(err, "error getting parameters from config")
+	}
+	params, ok := tfParams.(*TFLiteConfig)
+	if !ok {
+		err := utils.NewUnexpectedTypeError(params, tfParams)
+		return &Model{}, errors.Wrapf(err, "can not register tflite cpu model %s", conf.Name)
+	}
+	return CreateTFLiteCPUModel(ctx, params)
+}
+
+// CreateTFLiteCPUModel is a constructor that builds a tflite cpu implementation of the MLMS.
+func CreateTFLiteCPUModel(ctx context.Context, params *TFLiteConfig) (*Model, error) {
+	// Given those params, add the model
+	model, err := addTFLiteModel(ctx, params.ModelPath, &params.NumThreads)
+	if err != nil {
+		return &Model{}, errors.Wrapf(err, "could not add model from location %s", params.ModelPath)
+	}
+	return &Model{attrs: *params, model: model}, nil
+}
+
+// Infer takes the input map, finds the image (labeled 'image'), and uses the inference
+// package to return the result from the tflite cpu model as a map.
+func (m *Model) Infer(ctx context.Context, input map[string]interface{}) (map[string]interface{}, error) {
+	_, span := trace.StartSpan(ctx, "service::mlmodel::tflite_cpu::Infer")
+	defer span.End()
+
+	outMap := make(map[string]interface{})
+
+	f := func(key string) bool { _, ok := input[key]; return ok }
+	doInfer := func(key string) (map[string]interface{}, error) {
+		outTensors, err := m.model.Infer(input[key])
+		if err != nil {
+			return nil, errors.Wrap(err, "couldn't infer from model")
+		}
+		// Fill in the output map with the names from metadata if u have them
+		// Otherwise, do output1, output2, etc.
+		for i := 0; i < len(m.model.Info.OutputTensorTypes); i++ {
+			// name := m.metadata.Outputs[i].Name
+			// if name == "" {
+			if m.metadata != nil {
+				outMap[m.metadata.Outputs[i].Name] = outTensors[i]
+			} else {
+				outMap["output"+strconv.Itoa(i)] = outTensors[i]
+			}
+		}
+		return outMap, nil
+	}
+
+	switch {
+	case f("image"):
+		return doInfer("image")
+	case f("text"):
+		return doInfer("text")
+	default:
+		return doInfer("input")
+	}
```

Comment:
All of these do the same thing - it's just the key that is different. If you have the metadata, instead of a switch, we need to use the key string that matches what the metadata says it should be. If there is no metadata, then just Infer on whatever the one tensor is.